lib/puppet/graph/relationship_graph.rb

```diff
@@ -244,6 +244,7 @@ def replace_containers_with_anchors(catalog)
     containers.each { |x|
       admissible[x] = whit_class.new(:name => "admissible_#{x.ref}", :catalog => catalog)
       completed[x]  = whit_class.new(:name => "completed_#{x.ref}",  :catalog => catalog)
+      admissible[x].set_tags(x)
```

Comment:
I think we actually want to set the tags on both the completed_ and admissible_ whits. The reason being, if we don't include the completed_ whit, we still get a debug message about Class[Foo] not being tagged with the tags of the originating resource, from when we evaluate whether it's completed or not and if dependent resources should be refreshed.

ie.,
in the case described in the original ticket, only setting them on admissible_ will result in output like this:
```ruby
Debug: Exec[test](provider=posix): Executing '/usr/bin/true'
Debug: Executing: '/usr/bin/true'
Notice: /Stage[main]/Test/Exec[test]/returns: executed successfully
Debug: /Stage[main]/Test/Exec[test]: The container Class[Test] will propagate my refresh event
Info: /Stage[main]/Test/Exec[test]: Scheduling refresh of Class[Foo]
Debug: Class[Test]: The container Stage[main] will propagate my refresh event
Info: Class[Foo]: Scheduling refresh of Exec[notify test]
Debug: Exec[notify test](provider=posix): Executing '/usr/bin/true'
Debug: Executing: '/usr/bin/true'
Notice: /Stage[main]/Foo/Exec[notify test]: Triggered 'refresh' from 1 event
Debug: /Stage[main]/Foo/Exec[notify test]: The container Class[Foo] will propagate my refresh event
Debug: Class[Foo]: Not tagged with test
Debug: Class[Foo]: Resource is being skipped, unscheduling all events
Info: Class[Foo]: Unscheduling all events on Class[Foo]
```
which confirms that the dependent exec fired, but is a little confusing since afterwards it claims that resource shouldn't have been evaluated.

Can you think of any reason why we wouldn't want to set the tags on the completed_ whit? I think it should be safe because anything dependent on it will still be evaluated for their own tags, but curious what you think.

Reply:
I noticed that when I was verifying this and it definitely confused me. I had to just conclude it worked somehow even without being tagged. If there are no major consequences, I'm definitely +1 to fixing it for clarity.